src/main/java/org/threadly/litesockets/tcp/SSLProcessor.java

```diff
@@ -0,0 +1,228 @@
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLEngineResult.HandshakeStatus;
+import javax.net.ssl.SSLEngineResult.Status;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLSession;
+
+import org.threadly.concurrent.future.ListenableFuture;
+import org.threadly.concurrent.future.SettableListenableFuture;
+import org.threadly.litesockets.Client;
+import org.threadly.litesockets.utils.MergedByteBuffers;
+import org.threadly.litesockets.utils.TransactionalByteBuffers;
+
+/**
+ * This is a generic SSLClient that can be used to create an encrypted connection to a server.
+ * By default it will not check the servers certs and will only do TLS connections.
+ * 
+ * @author lwahlmeier
+ */
+class SSLProcessor {
+  //Not sure why but android needs this extra buffer
+  //as the getApp/Packet buffers are not right.
+  public static final int EXTRA_BUFFER_AMOUNT = 50;
+  
+  /**
+   * This is how much extra buffer we allocate for read events so we do quite
+   * as much allocating.  We always have to have at least 1 buffer the size SSLEngine says
+   * we need, but might use very little of it.  This way we can allocate once and get multiple
+   * read events with it before we have to throw away allocated bytes an allocate more. 
+   * 
+   */
+  public static final int PREALLOCATE_BUFFER_MULTIPLIER = 3;
+
+  private final AtomicBoolean finishedHandshake = new AtomicBoolean(false); 
+  private final AtomicBoolean startedHandshake = new AtomicBoolean(false);
+  private final SettableListenableFuture<SSLSession> handshakeFuture = new SettableListenableFuture<SSLSession>(false);
+  private final TransactionalByteBuffers encryptedReadBuffers = new TransactionalByteBuffers();
+  private final SSLEngine ssle;
+  private final Client client;
+  private ByteBuffer writeBuffer;
+  private ByteBuffer decryptedReadBuffer;
+
+  public SSLProcessor(final Client client, final SSLEngine ssle) {
+    this.client = client;
+    this.ssle = ssle;
+  }
+  
+  public boolean handShakeStarted() {
+    return startedHandshake.get();
+  }
+
+  /**
+   * This lets you know if the connection is currently encrypted or not.
+   * @return true if the connection is encrypted false if not.
+   */
+  public boolean isEncrypted() {
+    return (startedHandshake.get() && !ssle.getSession().getProtocol().equals("NONE"));
+  }
+
+  /**
+   * <p>If doHandshake was set to false in the constructor you can start the handshake by calling this method.
+   * The client will not start the handshake till its added to a SocketExecuter.  The future allows you to know
+   * when the handshake has finished if if there was an error.  While the handshake is processing all writes to the 
+   * socket will queue.</p>
+   * 
+   * @return A ListenableFuture.  If a result was given it succeeded, if there is an error it failed.  The connection is closed on failures.
+   */
+  public ListenableFuture<SSLSession> doHandShake() {
+    if(startedHandshake.compareAndSet(false, true)) {
```

Comment:
doHandshake is ran by startSSL() in the tcpClient, which can really be done by any thread the implementor chooses.  The thought here is if for some reason someone had multipule threads call it we would only ever allow one through but return the same future to both.